### PR TITLE
Missing corp api functions

### LINKED ISSUE #2981: [Design] Make corp bulk purchasing available at game start
Purchasing specific quantities of goods for a corp is frustrating for a number of reasons:
* You need to divide the amount you want by 10 to get the actual quantity you will get
* You then need to cancel the order or you will buy twice (or more) as much as you wanted
* If you are buying programmatically then if for any reason your script halts during the purchasing phase you will completely fill your warehouses
* You cannot purchase goods you are using to produce products while smart supply is active (ie you cannot purchase hardware for a software division without disabling smart supply)

The Bulk Purchasing research partially fixes some of this, but it has a number of other issues
* It costs research points, meaning it typically isn't available until after you no longer really need it
* It has no API support
* It can't be used while smart supply is purchasing the given material
* Unlike normal purchasing it requires you to have sufficient funds to buy the products listed

Overall, this means that Bulk Purchasing largely does not have any place unless you are handling your corp nearly completely by hand, and even then is unlikely to be used when you could just suffer to get a slight benefit of more research points.

I would like to make the following changes to resolve these issues:
* Remove the Bulk Purchasing research and make bulk purchasing available for free
* Change bulk purchasing to happen during the normal buy phase. The logic for that will be as follows:
  * First, make recurring material purchases
  * Then, buy as much of the bulk purchasing order as possible
  * Then, remove the purchased amount of material from the outstanding bulk purchase amount
* Make the purchase material dialog available regardless of whether Smart Supply is enabled or not. If Smart Supply is purchasing that material, only the buy material per second text field will be disabled.
* Add a new Corporation API method buyBulkMaterial, which will set the amount of the material to be purchased in bulk
* Add two new fields to the Material object returned from getMaterial, buy (the continuous amount of material purchased) and buyBulk (the amount of material to be purchased in bulk)

I also have the following outstanding questions for this design:
* Should bulk purchasing be renamed?
* Should this ability still be locked behind corp funds (e.g. $5b)?
* Should bulk purchasing orders show up in the material change per second UI?

### PR DESCRIPTION
**Pull request by phyzical**
(not sure what happened but this was part of @pigalot 's set of changes in the original big pr must have been lost in a force push or merge?)

fixes #2981

* adds api access to buyBackShares, sell shares and bulk purchase
* fixes a bug with make product when you cant for that industry
```
/** @param {NS} ns **/
export async function main(ns) {
    // please use dev menu to go to BN.3
    // give yourself a pile of cash
    //give corp a bunch of research
    ns.corporation.createCorporation("a", false)
    ns.corporation.goPublic(500000000)
    let shares = ns.corporation.getCorporation().numShares
    ns.corporation.buyBackShares(100000)
    if (shares + 100000 != ns.corporation.getCorporation().numShares) {
        throw new Error("didnt sell shares")
    }
    ns.corporation.sellShares(50000000)
    if (shares - 50000000 + 100000 != ns.corporation.getCorporation().numShares) {
        throw new Error("didnt buyback shares")
    }
    ns.corporation.expandIndustry("Software", "a")
    ns.corporation.expandCity("a", "Sector-12")
    ns.corporation.research("a", "Bulk Purchasing")
    let materials = ns.corporation.getMaterial("a", "Sector-12", "RealEstate").qty
    ns.corporation.bulkPurchase("a", "Sector-12", "RealEstate", 1000)
    if (materials + 1000 != ns.corporation.getMaterial("a", "Sector-12", "RealEstate").qty) {
        throw new Error("didnt buy materials")
    }
}
```

(will add a gif soon)

### CODE REVIEW

File: src/Corporation/ui/BuybackSharesModal.tsx [hunk 1]
```diff
@@ -36,22 +38,14 @@ export function BuybackSharesModal(props: IProps): React.ReactElement {
...
-      console.warn("Converting to number now");
-      const res = corp.issuedShares;
-      if (isNaN(res)) {
-        corp.issuedShares = 0;
-      } else {
-        corp.issuedShares = res;
-      }
+    try {
+      BuyBackShares(corp, player, shares)
+      props.onClose();
```
Comment (Mawootad) [inline src/Corporation/ui/BuybackSharesModal.tsx:43]: These should probably be outside of the try catch.

File: src/Corporation/ui/PurchaseMaterialModal.tsx [hunk 2]
```diff
@@ -54,32 +54,16 @@ interface IBPProps {
...
-      if (corp.funds >= cost) {
-        corp.funds = corp.funds - cost;
-        props.mat.qty += amount;
-      } else {
-        dialogBoxCreate(`You cannot afford this purchase.`);
-        return;
-      }
+    try {
+      BulkPurchase(corp, props.warehouse, props.mat, parseFloat(buyAmt));
       props.onClose();
```
Comment (Mawootad) [inline src/Corporation/ui/PurchaseMaterialModal.tsx:64]: Should there be a rerender here?

File: src/NetscriptFunctions/Corporation.ts [hunk 3]
```diff
@@ -431,6 +435,18 @@ export function NetscriptCorporation(
       const material = getMaterial(divisionName, cityName, materialName);
       BuyMaterial(material, amt);
     },
+    bulkPurchase: function (adivisionName: any, acityName: any, amaterialName: any, aamt: any): void {
+      checkAccess("bulkPurchase", 7);
+      const divisionName = helper.string("bulkPurchase", "divisionName", adivisionName);
+      if (!hasResearched(getDivision(adivisionName), "Bulk Purchasing")) throw new Error(`You have not researched Bulk Purchase in ${adivisionName}`)
```
Comment (Mawootad) [inline src/NetscriptFunctions/Corporation.ts:441]: ${adivisionName}

Should this be using ${divisionName} instead?

**Review by Mawootad — CHANGES_REQUESTED** [review]

**Comment by phyzical** [comment]
@Mawootad thx for having a look, will review the comments later today

File: src/Corporation/Actions.ts [hunk 4]
```diff
@@ -245,6 +247,58 @@ export function BuyMaterial(material: Material, amt: number): void {
...
+  player.gainMoney(profit, "corporation");
+  return profit;
+}
+
+export function BuyBackShares(corporation: ICorporation, player: IPlayer, numShares: number): boolean {
+  if (isNaN(numShares)) throw new Error("Invalid value for number of shares");
+  if (numShares < 0) throw new Error("Invalid value for number of shares");
+  if (numShares > corporation.issuedShares) throw new Error("You don't have that many shares to buy!");
+  if (!corporation.public) throw new Error("You haven't gone public!");
+  if (corporation.funds < (numShares * corporation.sharePrice)) throw new Error("You cant afford that many shares!");
```
[the remark this answers is not in the document]
Comment (phyzical) [reply src/Corporation/Actions.ts:293]: but why is the share price 10% higher? is the tax missing or something?

[the remark this answers is not in the document]
Comment (phyzical) [reply src/Corporation/Actions.ts:293]: oh yep buyback price is 1.1 ill just move it up

File: src/Corporation/ui/BuybackSharesModal.tsx (hunk 1 above)
Comment (phyzical) [reply src/Corporation/ui/BuybackSharesModal.tsx:43]: hmm possible ill see what happens in the "onclose" but i think your right 👍

File: src/NetscriptFunctions/Corporation.ts (hunk 3 above)
Comment (phyzical) [reply src/NetscriptFunctions/Corporation.ts:441]: good spot

**Pull request merged**